File: day3/ruby/part_2.rb
memory = File.read('input.txt')
regex = /mul\((\d+),(\d+)\)|do\(\)|don't\(\)/
result = 0
enabled = true


memory.scan(regex) do |match|
  if match[0] == 'do()'
    enabled = true
  elsif match[0] == "don't()"
    enabled = false
  elsif enabled && match[0].start_with?('mul')
    x, y = match[0].match(/\d+/g).map(&:to_i)
    result += x * y
  end
end

puts result

File: day3/python/part_2.py
import re

with open("input.txt") as f:
    memory = f.read()

enabled = True
result = 0

for instr in re.findall(r"mul\(\d+,\d+\)|do\(\)|don't\(\)", memory):
    if instr == "do()":
        enabled = True
    elif instr == "don't()":
        enabled = False
    elif enabled:
        x, y = map(int, re.findall(r"\d+", instr))
        result += x * y

print(result)
